pkg/bloomgateway/worker.go

```diff
@@ -0,0 +1,240 @@
+// the tasks.
+// It is responsible for multiplexing tasks so they can be processes in a more
+// efficient way.
+type worker struct {
+	services.Service
+
+	id      string
+	cfg     workerConfig
+	queue   *queue.RequestQueue
+	store   bloomshipper.Store
+	tasks   *pendingTasks
+	logger  log.Logger
+	metrics *workerMetrics
+}
+
+func newWorker(id string, cfg workerConfig, queue *queue.RequestQueue, store bloomshipper.Store, tasks *pendingTasks, logger log.Logger, metrics *workerMetrics) *worker {
+	w := &worker{
+		id:      id,
+		cfg:     cfg,
+		queue:   queue,
+		store:   store,
+		tasks:   tasks,
+		logger:  log.With(logger, "worker", id),
+		metrics: metrics,
+	}
+	w.Service = services.NewBasicService(w.starting, w.running, w.stopping).WithName(id)
+	return w
+}
+
+func (w *worker) starting(_ context.Context) error {
+	level.Debug(w.logger).Log("msg", "starting worker")
+	w.queue.RegisterConsumerConnection(w.id)
+	return nil
+}
+
+func (w *worker) running(ctx context.Context) error {
+	idx := queue.StartIndexWithLocalQueue
+
+	requests := make([]v1.Request, 0, 128)
+	fingerprints := make([]uint64, 0, 1024)
+
+	for ctx.Err() == nil {
+		taskCtx := context.Background()
+		dequeueStart := time.Now()
+		items, newIdx, err := w.queue.DequeueMany(taskCtx, idx, w.id, w.cfg.maxItems, w.cfg.maxWaitTime)
+		w.metrics.dequeueWaitTime.WithLabelValues(w.id).Observe(time.Since(dequeueStart).Seconds())
+		if err != nil {
+			// We only return an error if the queue is stopped and dequeuing did not yield any items
+			if err == queue.ErrStopped && len(items) == 0 {
+				return err
+			}
+			w.metrics.dequeueErrors.WithLabelValues(w.id).Inc()
+			level.Error(w.logger).Log("msg", "failed to dequeue tasks", "err", err, "items", len(items))
+		}
+		idx = newIdx
+
+		if len(items) == 0 {
+			continue
+		}
+		w.metrics.dequeuedTasks.WithLabelValues(w.id).Add(float64(len(items)))
+
+		tasksPerDay := make(map[time.Time][]Task)
+
+		for _, item := range items {
+			task, ok := item.(Task)
+			if !ok {
+				// This really should never happen, because only the bloom gateway itself can enqueue tasks.
+				return errors.Errorf("failed to cast dequeued item to Task: %v", item)
+			}
+			level.Debug(w.logger).Log("msg", "dequeued task", "task", task.ID)
+			w.tasks.Delete(task.ID)
+
+			fromDay, throughDay := task.Bounds()
+
+			if fromDay.Equal(throughDay) {
+				tasksPerDay[fromDay] = append(tasksPerDay[fromDay], task)
+			} else {
+				// split task into separate tasks per day
+				for i := fromDay; i.Before(throughDay); i = i.Add(24 * time.Hour) {
+					r := filterRequestForDay(task.Request, i)
+					t := task.CopyWithRequest(r)
+					tasksPerDay[i] = append(tasksPerDay[i], t)
+				}
+			}
+		}
+
+		for day, tasks := range tasksPerDay {
+			logger := log.With(w.logger, "day", day)
+			level.Debug(logger).Log("msg", "process tasks", "tasks", len(tasks))
+
+			it := newTaskMergeIterator(tasks...)
+
+			fingerprints = fingerprints[:0]
```

Comment:
nit: I don't think we need another list of all the fingerprints here. We could instead build a function which found the intersecting blocks for a list of fingerprints from the original requests.

Simplified below,

```go
func OverlappingBlocksForRequests(reqs [][]model.Fingerprint, blocks []Block) []Block
```

We could binary search over the fp lists, comparing them to blocks rather than iterate over every fp directly (can easily be `n=millions`).

Reply:
I've gone ahead and put together something like this here: https://github.com/grafana/loki/pull/11237

Reply:
You'd still need the list of fingerprints to calculating the overlapping blocks. Or do I miss something?